test/src/test/java/org/corfudb/integration/CorfuReplicationUpgradeIT.java

```diff
@@ -110,14 +94,9 @@ public void testLRNotStartedUntilRollingUpgradeCompletes() throws Exception {
         defaultAdapterForUpgrade = new DefaultAdapterForUpgrade(sourceRuntime);
         defaultAdapterForUpgrade.startRollingUpgrade();
 
-        // Two updates are expected onStart of snapshot sync and onEnd.
-        CountDownLatch latchSnapshotSyncPlugin = new CountDownLatch(2);
-        snapshotSyncPluginListener = new SnapshotSyncPluginListener(latchSnapshotSyncPlugin);
-        subscribeToSnapshotSyncPluginTable(snapshotSyncPluginListener);
-
-        CountDownLatch statusUpdateLatch = new CountDownLatch(NUM_SNAPSHOT_SYNC_UPDATES_ON_SINK_STATUS_TABLE);
-        sinkListener = new ReplicationStatusListener(statusUpdateLatch, false);
-        corfuStoreSink.subscribeListener(sinkListener, CORFU_SYSTEM_NAMESPACE, LR_STATUS_STREAM_TAG);
+        CountDownLatch statusUpdateLatch = new CountDownLatch(1);
+        sourceStatusListener = new ReplicationStatusListener(statusUpdateLatch, true);
+        corfuStoreSource.subscribeListener(sourceStatusListener, CORFU_SYSTEM_NAMESPACE, LR_STATUS_STREAM_TAG);
```

Comment:
Can we sort of combine these 2 checks? i.e. the Source status table COMPLETE check and Sink side snapshot plugin start & end check?

I have 2 concerns for only relying on Source status table:

One is it might lead to false positive results, like in this test case, if some errors happened w.r.t. the forced snapshot sync after rolling upgrade happened, the negotiation result (which is continue log entry sync) could still make all the following check pass.

Second one is there could be race condition between the forced snapshot sync (driven by FSM) and the negotiation (driven by runtimeFSM) - if negotiation finishes first and FSM already transits into `InLogEntrySyncState`, Source status table will be marked with COMPLETE and the count down latch completes (before the forced snapshot sync even happens), which is undesired

Reply:
Thanks.
We cannot wait for start+end on the plugin because there can be 2 starts if snapshot sync restarts.  But the test can wait for end instead.

I see the point wrt the limitations of the check on only the Source.  Will modify it.